[AMDGPU] clang misreports HIP installation path as `/usr/local` when it is installed in `/usr`.

clang misreports HIP installation path as `/usr/local` when it is installed in `/usr` (in Gentoo).

Context: CMake calls `clang -v -print-targets` to get HIP/ROCm root: [CMakeDetermineHIPCompiler.cmake#L151](https://github.com/Kitware/CMake/blob/v3.28.1/Modules/CMakeDetermineHIPCompiler.cmake#L151)

```
$ clang -v -print-targets -print-rocm-search-dirs 
clang version 17.0.6
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/lib/llvm/17/bin
Configuration file: /etc/clang/x86_64-pc-linux-gnu-clang.cfg
System configuration file directory: /etc/clang
Selected GCC installation: /usr/lib/gcc/x86_64-pc-linux-gnu/13
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Selected multilib: .;@m64
ROCm installation search path: /usr/lib/llvm/17
ROCm installation search path: /usr/lib/llvm/17
ROCm installation search path: /usr/lib/llvm/17/bin/../../../../lib/clang/17
ROCm installation search path: /opt/rocm
ROCm installation search path: /usr/local
ROCm installation search path: /usr
Found HIP installation: /usr/local, version 5.7.31921
...
```

And fails with:
```
CMake Error at /usr/share/cmake/Modules/CMakeDetermineHIPCompiler.cmake:217 (message):
  The ROCm root directory:

   /usr/local

  does not contain the HIP runtime CMake package, expected at one of:

   /usr/local/lib/cmake/hip-lang/hip-lang-config.cmake
   /usr/local/lib64/cmake/hip-lang/hip-lang-config.cmake

Call Stack (most recent call first):
  Applications/bitonic_sort/CMakeLists.txt:37 (enable_language)
```

Meanwhile, `/usr/local` is empty ([docs](https://devmanual.gentoo.org/general-concepts/filesystem/index.html))
```sh
$ du /usr/local
0       /usr/local/bin
0       /usr/local/lib
0       /usr/local/sbin
0       /usr/local/lib64
0       /usr/local
```

In Gentoo HIP/ROCm is installed into `/usr`. Version file is installed into `/usr/share/hip/version`. HIP runtime is detected by 
https://github.com/llvm/llvm-project/blob/bbb3f9b1321176a012dac7d50da54b869a85e9bc/clang/lib/Driver/ToolChains/AMDGPU.cpp#L478-L492
It checks ROCm installation search paths, at some point it checks `/usr/local` (before `/usr`), then according to the code finds `/usr/local/../share/hip/version` and completes the search.

I examined clang code and I don't see any way to manipulate clang to report the correct path. There is `ROCm installation search path: /usr`, but it is: 1) checked after `/usr/local` and 2) checked by trying to find a file `/usr/../share/hip/version`, which does not exist.

I'm aware that it is possible to patch CMakeDetermineHIPCompiler.cmake or set env vars HIP_PATH, ROCM_PATH, etc, but I want to see correct path detection out-of-the-box without patches or extra environment variables (maybe with CMake option like `-DCLANG_ROCM_DIR=/usr` for Gentoo Clang ebuild maintainers).

Could you revisit ROCm detection code and provide any solution? Thanks!

--- a/clang/lib/Driver/ToolChains/AMDGPU.cpp
+++ b/clang/lib/Driver/ToolChains/AMDGPU.cpp
@@ -486,10 +486,16 @@ void RocmInstallationDetector::detectHIPRuntime() {
       return newpath;
     };
     // If HIP version file can be found and parsed, use HIP version from there.
-    for (const auto &VersionFilePath :
-         {Append(SharePath, "hip", "version"),
-          Append(ParentSharePath, "hip", "version"),
-          Append(BinPath, ".hipVersion")}) {
+    std::vector<SmallString<0>> VersionFilePaths = {
+        Append(SharePath, "hip", "version"),
+        InstallPath != D.SysRoot + "/usr/local"
+            ? Append(ParentSharePath, "hip", "version")
+            : SmallString<0>(),
+        Append(BinPath, ".hipVersion")};
+
+    for (const auto &VersionFilePath : VersionFilePaths) {
+      if (VersionFilePath.empty())
+        continue;
       llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> VersionFile =
           FS.getBufferForFile(VersionFilePath);
       if (!VersionFile)

--- a/clang/test/Driver/rocm-detect.hip
+++ b/clang/test/Driver/rocm-detect.hip
@@ -77,8 +77,18 @@
 // RUN:   --hip-path=%t/myhip --print-rocm-search-dirs %s 2>&1 \
 // RUN:   | FileCheck -check-prefixes=ROCM-ENV,HIP-PATH %s
 
+// Test detecting /usr directory.
+// RUN: rm -rf %t/*
+// RUN: cp -r %S/Inputs/rocm %t/usr
+// RUN: mkdir -p %t/usr/share/hip
+// RUN: mv %t/usr/bin/.hipVersion %t/usr/share/hip/version
+// RUN: mkdir -p %t/usr/local
+// RUN: %clang -### --target=x86_64-linux-gnu --offload-arch=gfx1010 --sysroot=%t \
+// RUN:   --print-rocm-search-dirs --hip-link %s 2>&1 \
+// RUN:   | FileCheck -check-prefixes=USR %s
+
 // Test detecting latest /opt/rocm-{release} directory.
-// RUN: rm -rf %t/opt
+// RUN: rm -rf %t/*
 // RUN: mkdir -p %t/opt
 // RUN: cp -r %S/Inputs/rocm %t/opt/rocm-3.9.0-1234
 // RUN: cp -r %S/Inputs/rocm %t/opt/rocm-3.10.0
@@ -129,6 +139,11 @@
 // ROCM-PATH: "-mlink-builtin-bitcode" "[[ROCM_PATH]]/amdgcn/bitcode/oclc_isa_version_1010.bc"
 // ROCM-PATH: "-idirafter" "[[ROCM_PATH]]/include"
 // ROCM-PATH: "-L[[ROCM_PATH]]/lib" {{.*}}"-lamdhip64"
+
+// USR: ROCm installation search path: [[ROCM_PATH:.*/usr$]]
+// USR: "-mlink-builtin-bitcode" "[[ROCM_PATH]]/amdgcn/bitcode/oclc_isa_version_1010.bc"
+// USR: "-idirafter" "[[ROCM_PATH]]/include"
+// USR: "-L[[ROCM_PATH]]/lib" {{.*}}"-lamdhip64"
 
 // ROCM-REL: ROCm installation search path: {{.*}}/opt/rocm
 // ROCM-REL: ROCm installation search path: {{.*}}/opt/rocm-3.10.0

